proposals/NNNN-swift-noasync.md

```diff
@@ -0,0 +1,329 @@
+
+## Introduction
+
+The Swift concurrency model allows tasks to resume on different threads from the
+one they were suspended on. For this reason, API that relies on thread-local
+storage, locks, mutexes, and semaphores, should not be used across suspension
+points.
+
+```swift
+func badAsyncFunc(_ mutex: UnsafeMutablePointer<pthread_mutex_t>, _ op : () async -> ()) async {
+  // ...
+  pthread_mutex_lock(mutex)
+  await op()
+  pthread_mutex_unlock(mutex) // Bad! May unlock on a different thread!
+  // ...
+}
+```
+
+The example above exhibits undefined behaviour if `badAsyncFunc` resumes on a
+different thread than the one it started on after running `op` since
+`pthread_mutex_unlock` must be called from the same thread that locked the
+mutex.
+
+We propose extending `@available` with a new `noasync` availability kind to
+indicate API that may not be used directly from asynchronous contexts.
+
+Swift evolution thread: [Pitch: Unavailability from asynchronous contexts](https://forums.swift.org/t/pitch-unavailability-from-asynchronous-contexts/53877)
+
+## Motivation
+
+The Swift concurrency model allows tasks to suspend and resume on different
+threads. While this behaviour allows higher utility of computational resources,
+there are some nasty pitfalls that can spring on an unsuspecting programmer. One
+such pitfall is the undefined behaviour from unlocking a `pthread_mutex_t` from
+a different thread than the thread that holds the lock, locking threads may
+easily cause unexpected deadlocks, and reading from and writing to thread-local
+storage across suspension points may result in unintended behaviour that is
+difficult to debug.
+
+## Proposed Solution
+
+We propose extending `@available` to accept a `noasync` availability kind.
+The `noasync` availability kind is applicable to most declarations, but is not
+allowed on destructors as those are not explicitly called and must be callable
+from anywhere.
+
+```swift
+@available(*, noasync)
+func doSomethingNefariousWithNoOtherOptions() { }
+
+@available(*, noasync, message: "use our other shnazzy API instead!")
+func doSomethingNefariousWithLocks() { }
+
+func asyncFun() async {
+  // Error: doSomethingNefariousWithNoOtherOptions is unavailable from
+  //        asynchronous contexts
+  doSomethingNefariousWithNoOtherOptions()
+
+  // Error: doSomethingNefariousWithLocks is unavailable from asynchronous
+  //        contexts; use our other shanzzy API instead!
+  doSomethingNefariousWithLocks()
+}
+```
+
+Using an annotated API from an asynchronous context results in an error.
+In certain cases, it is possible to use the API safely within an asynchronous
+context, but not across suspension points. The attribute is only checked in the
+immediate context, so an unavailable function call wrapped in a synchronous
+context will not emit the error. This allows functions to wrap the behaviour and
```

Comment:
```suggestion
context, but not across suspension points. The attribute only prevents uses where
the immediate context is `async`. So if an unavailable function is called from a 
synchronous function wrapper, that is not an error, despite the wrapper itself being
called from an async context. This allows functions to limit the `noasync` behaviour to
```

Reply:
Re-worded this paragraph a bit. Wasn't a huge fan of the flow.